### DISCUSSION TITLE
Earthlyfiles vs Dockerfiles in the context of ease of contribution

### DISCUSSION THREAD

**Discussion by EzraBrooks** [space-ros/space-ros#114] [General]
See #107 

We probably shouldn't use both and confuse contributors. Need to decide on one to make it less confusing

**Comment by xfiderek** [comment]
## Earthlyfiles vs Dockerfiles usage as of now


| Repo/Folder  | Type of file | Use case |
| ------------- | -------------------------------| ------------- |
|  [spaceros](https://github.com/space-ros/space-ros/blob/main/Earthfile)  | 🌎 Earthly | Generate vcs.repos file  |
| [docker/spaceros](https://github.com/space-ros/docker/blob/main/spaceros/Earthfile)  | 🌎 Earthly  | build spaceros docker image, run colcon tests, push image to registry  |
| [docker/moveit2](https://github.com/space-ros/docker/blob/main/moveit2/Dockerfile)  | 🐋 Docker  | build moveit docker image from spaceros image _(note that there is no action for running tests and pushing to registry)_  |
| [docker/space_robots](https://github.com/space-ros/docker/tree/main/space_robots)  | 🐋 Docker  | build space_robots docker image from spaceros and moveit image _(note that there is no action for running tests and pushing to registry)_  |
| [rtems, renode, zynq](https://github.com/space-ros/docker) |  3 x 🐋 Docker  | Build docker images for other demos. Note that these  are not using spaceros image  |


**Looks like main use cases for earthly today are as follows:**
- run vcs generation scripts and save generated ‘vcs.repos’ to local
- run colcon test on the newest code and save test results to local
- build & publish docker images
- *Anything i missed here?*

## Earthly pros and cons

**Pros of Earthly (i.e. why we should keep it)**
- It is already in place for the most frequently used workflows 😀
- Earthly scripts are reproducible on every machine. To ensure the same without earthly we would need to create a docker image for every action, e.g. [running vcs scripts](https://github.com/space-ros/space-ros/tree/main/scripts)
- High reusability of workflows / harmonization. We can import tasks from other Earthfiles and reuse them in downstream workflows

**Cons of Earthly (i.e. why we should get rid of it)**
- Additional learning curve for most contributors
- bash scripts / Makefiles are more standard

## Possible replacement
Every earthly workflow could be replaced by a single Dockerfile with bunch of scripts that are copied to the container
For example we would replace the [docker/spaceros/Earthfile](https://github.com/space-ros/docker/tree/main/spaceros) with the following files:
```
├── spaceros
│   ├── scripts - folder with scripts that are copied to the docker container
│   │   ├── entrypoint.sh - default entrypoint
│   │   ├── run_tests.sh - script for running tests (colcon test)
│   ├── Dockerfile -  image definition (set env variables, pull vcs repos, colcon build)
│   ├── build.sh - builds docker image
│   ├── run.sh - runs docker container with default entrypoint
│   ├── test.sh - runs docker container with scripts/run_tests.sh and binds folder where results will be stored
```
To pass the output (e.g. test results) back to host we could simply use volumes defined in shell scripts.

The only problem that I have with this approach is that I am not sure what to do with [vcs generation scripts](https://github.com/space-ros/space-ros/blob/main/Earthfile). To ensure that .vcs can be generated on any machine, we would need to create Dockerfile as well. Where we would store this Dockerfile? in `docker` repo or in `spaceros` repo? 
If we store it in `docker` repo then generation of new vcs file and spaceros image will be a bit clunky (going back and forth between `docker` repo and `spaceros` repo). 
If we store it in `spaceros` repo, then `docker` repository will be a bit confusing, as Dockerfile will also be present in `spaceros` repo. 
**Any idea how to solve it nicely?**

## What if we decide to keep Earthly
There is bunch of cleanups / fixes that we need to perform: https://github.com/space-ros/docker/issues/88,
 but those would need to be fixed in shell scripts anyway if we decide to move away. 

Apart from that we would need to replace Dockerfiles in `docker` repo with Earthfiles (e.g. for moveit etc)

## My take
In my opinion, Earthly is a nice tool and as long as we build good quickstart page for out contributors, we should be fine with keeping it, as learning curve is not that high. Certainly, bash scripts and makefiles are more standard tools, but give us less harmonization potential and make the code more clunky. If we have a good template for using Earthfiles in the project, I think it should be ok :). Also it would minimize rework on our side and help us focus on other problems, as I think migration from Docker to Earthly will be easier than other way around.

Personally, I have never used Earthly prior to contributing to this project and I found it quite easy to learn (I mean, my first task is about Earthly). I find the tool quite good for our requirements, especially for things like vcs generation. If we want to lower learning curve further, we could create a page in our docs about earthly and how to use it.

Also, last release was 2 weeks ago and  the number of stars on github is steadily growing, which indicates that the project is in good shape and will continue to evolve:

<p align="center">
<img src="https://github.com/space-ros/space-ros/assets/57559956/d60f6da0-cca8-4a9b-9a94-809667a5036e" width="450" height="300">
</p>

**Comment by Bckempa** [comment]
CI is another aspect of this. We have some issues (https://github.com/space-ros/space-ros/issues/80, https://github.com/space-ros/space-ros/issues/90) that will add CI checks, nominally via GitHub Actions. Either we'd need to pick an external workflow provider (which introduces a supply-chain risk if the workflow is modified), or vendor in our own which increases the maintenance burden and locks Space ROS further into the GitHub platform.

Alternatively, we could implement them as actions in the Earthly files. While we'd still incur the additional maintenance, instead of maintaining a configuration specifically for GHA, the checks would use the same tech stack as our builds (reducing the number of platforms we need to understand to make changes) and can be run locally as part of the dev process without adding additional scripts that parrot the GHA workflow.

**Comment by Mechazo11** [comment]
@EzraBrooks,  @Bckempa I know this discussion was opened a while back but I wanted to chip in my own experience. I participated in NASA's space ros competition this year.  Docker was the go to choice for the submission and for most intents and purpose, it made the submission process very smooth. If I am not mistaken, using Docker we can even replicate complex errors (I am in the process of adding an Issue about a potential bug in base space-ros image).

Hence, my questions is, how Earthly is really helping Space ROS? The reason I ask is because only the base image build uses Earthly while all the other images (nav2, moveit2, space_robots_demo) are using Docker.

Apologies if I am not understanding the merits of Earthly but I agree with @EzraBrooks, maybe a separate tutorial that goes in to Earthly a bit more would be very helpful. But what really confused me was why Earthly was needed specifically to build the base image when all other images we have built was using Docker.

**Reply by Bckempa** [reply]
The reason Earthly is used for the base image, and why this approach has won over a few skeptics on the team (myself included) comes down to a pretty basic cost-benefit analysis: if you are hacking on the base image build, the more advanced Earthly caching can massively improve iteration time, but developer using Space ROS should never need to run it, you can pull a base image made by CI from Docker Hub and build all your applications on top using vanilla Docker, just as we demonstrate with all the demo images. Put another way, it improves things for maintainers but only maintainers should need to interface with it.

I am aware that some tutorials walk though the earthly build process. In my opinion (and my opinion alone, I'm not speaking for anyone else) we should make it clear that these are instructions for maintaining Space ROS itself and have a separate tutorial for building on those base images instead. Part of the reason those have become conflated is those docs were originally written when only maintainers were developing the project so the definition of who a "user" of Space ROS is has shifted; but, we also dont' want to remove them entirely not just because they are useful for on-boarding new maintainers, but also because flight projects adopting Space ROS will want to do their own source builds through their validated tool chain and must be able to replicate the process (or modify it as needed).

TL;DR: I think this is a communication issue, you shouldn't need to run Earthly to use Space ROS and we can and should make that more clear in the docs.

**Reply by Mechazo11** [reply]
@Bckempa thank you very much for the detailed reply. I agree with your analysis those of us who would primarily be be building on top of base space_ros shouldn't need to worry about Earthly at all.

**Reply by asimonov** [reply]
I agree with @Bckempa. I was sceptical at first about Earthly, but it was easy to learn and it does add value by making build less verbose. 

As with any new tool it also adds gotchas, as we have found while messing with Earthly caching strategy. But overall it is nice tool to use.

**Comment by DionisisGiannar** [comment]
Hello, space-ros community,

I am new in the space-ros. I would like to learn if the **osfr/space-ros** docker image pulled from Docker Hub and the image build from **Earthly** with Github steps, are the same or not.

Taking into account that more than the half of the repositories are from Docker Hub Image and the others from Earthly, which approach would you recommend: building the image using Earthly or pulling the pre-built image from Docker Hub?

**Reply by Bckempa** [reply]
The images pulled from Docker Hub are the release versions of SpaceROS you should build your applications on top of. You really only need to be doing the Earthly build if you're developing changes to the SpaceROS distribution itself.

**Reply by DionisisGiannar** [reply]
@Bckempa Thanks for applying! Now it's clearer to me, I will keep that in mind.

**Reply by asimonov** [reply]
Just want to clarify that Earhtly is just a tool that produces docker images, not some other format. They are pushed to docker hub.

**Comment by salvioli** [comment]
The Earthly project had some recent important developments. I think we should keep an eye on this to correctly inform this discussion.

https://github.com/earthly/earthly/issues/4313#issuecomment-2590566145

I am not sure exactly about the state of the conversation besides this discussion so I apologize if I am a bit late to the party or if this has already been reported/discussed.

**Reply by Bckempa** [reply]
Thanks for the heads-up! I've added this to the next [technical meeting agenda](https://github.com/space-ros/space-ros/discussions/247).

**Comment by salvioli** [comment]
New updates from Earthly [here](https://earthly.dev/blog/shutting-down-earthfiles-cloud/).

[This can also be interesting](
https://dagger.io/blog/earthly-to-dagger-migration?utm_campaign=Earthly-Migration&utm_medium=blog&utm_source=earthly-blog)

**Reply by asimonov** [reply]
"We are also ending active maintenance of the Earthly open-source project.

We are supporting the community’s efforts to self-organize a fork, and we encourage those interested to get involved."

:(

**Reply by Bckempa** [reply]
Added to the [technical meeting agenda](https://github.com/space-ros/space-ros/discussions/265#discussioncomment-12886528). Thanks for keeping us informed!
